### PR TITLE
Tejaswis/xms pageable for get list calls

### PR DESCRIPTION
**Pull request by tejaswiMinnu**
OAPI009: Always add x-ms-pageable to LIST calls
x-ms-pageable indicates that the results of a GET call may be paginated. All LIST APIs (a.k.a collection GETs) should include this annotation.

### CODE REVIEW

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft.azure/openapi-validator-rulesets",
+      "comment": "Added XmsPageableForGetListCalls (RPC-Get-V1-13) rule to ensure that all GET LIST APIs contain the ‘x-ms-pageable’ property",
```
Comment (AkhilaIlla) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json:5]: >Added [](http://example.com/codeflow?start=18&length=5)

Should be Add.
Thats the mandate it seems

Comment (AkhilaIlla) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json:5]: >LIST [](http://example.com/codeflow?start=95&length=4)

can be lower case?

File: docs/xms-pageable-for-get-list-calls.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+For all LIST APIs (a.k.a collection GETs), it is important to include the `x-ms-pageable` property.
+
+## Description
+
+`x-ms-pageable` indicates that the results of a GET call may be paginated. All LIST APIs (a.k.a collection GETs) should include this annotation.
+
+## How to fix the violation
+
+Please include `x-ms-pageable` property to the collection GET calls.
```
Comment (AkhilaIlla) [inline docs/xms-pageable-for-get-list-calls.md:25]: >property [](http://example.com/codeflow?start=31&length=8)

Term "property" is ambiguous, I mistook it be defined within properties.
We need to consider using a different term?

File: packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  if (swaggerObj === null) {
+    return []
+  }
+  const path = paths.path || []
+
+  if (swaggerObj["x-ms-pageable"]) return []
+  else
+    return [
+      {
+        message: "For all LIST APIs (a.k.a collection GETs), it is important to include the `x-ms-pageable` property.",
```
Comment (AkhilaIlla) [inline packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts:11]: >it is important to include the `x-ms-pageable` property. [](http://example.com/codeflow?start=61&length=56)

`x-ms-pageable' MUST be included?

File: packages/rulesets/src/spectral/test/xms-pageable-for-get-list-calls.test.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+beforeAll(async () => {
+  linter = await linterForRule("XmsPageableForGetListCalls")
+  return linter
+})
+
+test("XmsPageableForGetListCalls should find errors if x-ms-pagebale property is not defined.", () => {
+  const oasDoc = {
+    swagger: "2.0",
+    paths: {
+      "/{resourceUri}/providers/Microsoft.ConnectedVMwarevSphere/virtualMachine": {
```
Comment (AkhilaIlla) [inline packages/rulesets/src/spectral/test/xms-pageable-for-get-list-calls.test.ts:17]: >/{resourceUri}/providers/Microsoft.ConnectedVMwarevSphere/virtualMachine [](http://example.com/codeflow?start=7&length=72)

can we even add examples for regular and nested resourceTypes in addition to the extensionResourceTypes

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json (hunk 1 above)
Comment (rkmanda) [reply common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json:5]: Not a mandate but a best practice to use simple present tense.

Comment (rkmanda) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json:5]: >XmsPageableForGetListCalls [](http://example.com/codeflow?start=24&length=26)

XmsPageableForListCalls

File: docs/xms-pageable-for-get-list-calls.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+- RPC-Get-V1-13
+
+## Output Message
+
+For all LIST APIs (a.k.a collection GETs), it is important to include the `x-ms-pageable` property.
+
+## Description
+
+`x-ms-pageable` indicates that the results of a GET call may be paginated. All LIST APIs (a.k.a collection GETs) should include this annotation.
```
Comment (rkmanda) [inline docs/xms-pageable-for-get-list-calls.md:21]: >All LIST APIs (a.k.a collection GETs) should include this annotation. [](http://example.com/codeflow?start=75&length=69)

`x-ms-pageable` indicates that the results of a GET call may be paginated. All LIST APIs (a.k.a collection GETs) should include this annotation. Specifying this extension does not mean that the service must implement paging at runtime. The service may implement this when such a requirement arises. Specifying the extension upfront future proofs the API and avoids the breaking changes that may be flagged if the extension were to be introduced in future.

File: docs/xms-pageable-for-get-list-calls.md (hunk 2 above)
Comment (rkmanda) [reply docs/xms-pageable-for-get-list-calls.md:25]: extension is a good term to use

File: packages/rulesets/src/spectral/az-arm.ts [hunk 6]
```diff
@@ -351,6 +352,20 @@ const ruleset: any = {
       },
     },
 
+    // RPC Code: RPC-Get-V1-11
+    XmsPageableForGetListCalls: {
+      description: "For all LIST APIs (a.k.a collection GETs), it is important to include the `x-ms-pageable` property.",
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:357]: >property [](http://example.com/codeflow?start=110&length=8)

extension

Comment (rkmanda) [inline packages/rulesets/src/spectral/az-arm.ts:357]: >For all LIST APIs (a.k.a collection GETs), it is important to include the `x-ms-pageable` property [](http://example.com/codeflow?start=20&length=98)

`x-ms-pageable` extension must be specified for LIST APIs

File: packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+const xmsPageableForGetListCalls = (swaggerObj: any, _opts: any, paths: any) => {
+  if (swaggerObj === null) {
+    return []
+  }
+  const path = paths.path || []
+
+  if (swaggerObj["x-ms-pageable"]) return []
+  else
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts:8]: Lets also check that the definition of the extension is also specified correctly. 

The extension must have the 
nextLinkName property defined

**Review by rkmanda — CHANGES_REQUESTED** [review]
:clock1:

File: packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts (hunk 3 above)
Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts:11]: Yes

File: packages/rulesets/src/spectral/test/xms-pageable-for-get-list-calls.test.ts (hunk 4 above)
Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/test/xms-pageable-for-get-list-calls.test.ts:17]: I already added on line 38

File: packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts (hunk 7 above)
Comment (tejaswiMinnu) [reply packages/rulesets/src/spectral/functions/xms-pageable-for-get-list-calls.ts:8]: That is already covered in native rule

File: common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json (hunk 1 above)
Comment (tejaswiMinnu) [reply common/changes/@microsoft.azure/openapi-validator-rulesets/tejaswis-XmsPageableForGetListCalls_2023-09-21-23-34.json:5]: No, cause the rule focus only on LIST calls, so mentioning in Caps makes more sense.

File: docs/xms-pageable-for-list-calls.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+`x-ms-pageable` indicates that the results of a GET call may be paginated. All LIST APIs (a.k.a collection GETs) should include this annotation. Specifying this extension does not mean that the service must implement paging at runtime. The service may implement this when such a requirement arises. Specifying the extension upfront future proofs the API and avoids the breaking changes that may be flagged if the extension were to be introduced in future.
+
+## How to fix the violation
+
+Please include `x-ms-pageable` extension to the collection GET calls.
+
+### Valid Example
+
+```json
```
Comment (bdefoy) [inline docs/xms-pageable-for-list-calls.md:29]: It formats better with json5:
```suggestion
```json5
```

File: docs/xms-pageable-for-list-calls.md [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+      "nextLinkName": null,
+    },
+  }
+}
+...
+```
+
+### Invalid Example
+
+```json
```
Comment (bdefoy) [inline docs/xms-pageable-for-list-calls.md:92]:
```suggestion
```json5
```

**Review by bdefoy — APPROVED** [review]

**Review by rkmanda — APPROVED** [review]
:shipit:

**Pull request merged**